src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,595 @@
+		{
+			_xmlDocumentLocation = document;
+			_document = new XPathDocument (_xmlDocumentLocation);
+			_context = context;
+			Initialize ();
+		}
+
+		public DynamicallyAccessedMemberTypes GetFieldAnnotation (FieldDefinition field)
+		{
+			return _fields.TryGetValue (field, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetParameterAnnotation (MethodDefinition method, int index)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+				string paramName = method.Parameters[index].Name;
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == paramName)
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetPropertyAnnotation (PropertyDefinition property)
+		{
+			return _properties.TryGetValue (property, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetReturnParameterAnnotation (MethodDefinition method)
+		{
+			return _methods.TryGetValue (method, out var ann) ? ann.ReturnAnnotation : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetThisParameterAnnotation (MethodDefinition method)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == "this")
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		static DynamicallyAccessedMemberTypes GetMemberTypesForDynamicallyAccessedMemberAttribute (ArrayBuilder<Attribute> attributes, LinkContext _context, string _xmlDocumentLocation)
+		{
+			foreach (var attribute in attributes.ToArray ()) {
+				if (attribute.attributeName == "System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembers" && attribute.arguments.Count == 1) {
+					foreach (var argument in attribute.arguments.ToArray ()) {
+						if (argument == string.Empty)
+							break;
+						try {
+							return (DynamicallyAccessedMemberTypes) Enum.Parse (typeof (DynamicallyAccessedMemberTypes), argument);
+						} catch (ArgumentException) {
+							_context.LogMessage (MessageContainer.CreateWarningMessage ($"Could not parse argument '{argument}' specified in '{_xmlDocumentLocation}' as a DynamicallyAccessedMemberTypes", 2020));
+						}
+					}
+				}
+			}
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		void Initialize ()
+		{
+			XPathNavigator nav = _document.CreateNavigator ();
+
+			if (!nav.MoveToChild ("linker", string.Empty))
+				return;
+
+			try {
+				ProcessAssemblies (_context, nav.SelectChildren ("assembly", string.Empty));
+			} catch (Exception ex) when (!(ex is XmlResolutionException)) {
+				throw new XmlResolutionException (string.Format ("Failed to process XML description: {0}", _document), ex);
```

Comment:
Is this recoverable? I don't know how we treat the exception from the call site. If it's not recoverable I think this should use the new `LinkerFatalErrorException` from https://github.com/mono/linker/pull/1155.

Reply:
The PR is still not merged and therefore LinkerFatalErrorException does not exist yet, but it will use the error 1004 just like the ResolveXMLStep

Reply:
Also, the message is not correct